warp/tests/test_grad_struct_array.py

```
@@ -0,0 +1,74 @@
+        with tape:
+            wp.launch(kernel=pack_struct_array_kernel, dim=1, inputs=[x], outputs=[y])
+            wp.launch(kernel=loss_from_struct_array_kernel, dim=1, inputs=[y], outputs=[loss])
+
+        tape.backward(loss=loss)
+
+        # Check that gradients propagate correctly
+        assert_np_equal(y.grad.numpy()[0][0], 1.0, tol=1e-5)  # y.grad[0].a should be 1.0
+        assert_np_equal(x.grad.numpy()[0], 1.0, tol=1e-5)  # x.grad[0] should be 1.0 (not 0.0!)
+
+
+devices = get_test_devices()
```

Ruff (0.14.14)

[error] `get_test_devices` may be undefined, or defined from star imports

(F405)